### PR TITLE
[SofaCore] Clean MultiVecId

### PR DESCRIPTION
**Pull request by damienmarchal**
- move code from header to compilation unit
- remove the dependency to Data. 

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by damienmarchal** [comment]
[ci-build]

File: SofaKernel/modules/SofaCore/src/sofa/core/MultiVecId.h [hunk 1]
```diff
@@ -678,8 +606,16 @@ typedef TMultiVecId<V_MATDERIV, V_WRITE>     MultiMatrixDerivId;
 typedef TMultiVecId<V_ALL, V_READ>      ConstMultiVecId;
 typedef TMultiVecId<V_ALL, V_WRITE>          MultiVecId;
 
-} // namespace core
+//#if (SOFA_CORE_MULTIVECID_SKIP_EXTERN_TEMPLATE_DECLARATION == 1)
```
Comment (alxbilger) [inline SofaKernel/modules/SofaCore/src/sofa/core/MultiVecId.h]: Any reason keeping this commented code? If not, the macro declaration in the cpp file is also not needed.

**Comment by fredroy** [comment]
"windows10_vs-2017_options — Build failed. "
It smells like some _API keywords are missing 🧐

File: SofaKernel/modules/SofaCore/src/sofa/core/MultiVecId.h (hunk 1 above)
Comment (damienmarchal) [reply SofaKernel/modules/SofaCore/src/sofa/core/MultiVecId.h]: No :)

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests][with-regression-tests]

File: SofaKernel/modules/SofaCore/src/sofa/core/MultiVecId.h [hunk 2]
```diff
@@ -63,12 +61,11 @@ struct StateVecAccessor<DataTypes, V_COORD, V_WRITE>
...
 public:
     typedef TVecId<V_COORD, V_WRITE> MyVecId;
-    typedef Data<typename DataTypes::VecCoord> MyDataVec;
 
     StateVecAccessor(State<DataTypes>* state, const MyVecId& id) : state(state), id(id) {}
     operator MyVecId() const {  return id;  }
-    const MyDataVec* read()  const {  return state-> read(id);  }
-    MyDataVec* write() const {  return state->write(id);  }
+    auto read()  const {  return state-> read(id);  }
+    auto write() const {  return state->write(id);  }
```
Comment (guparan) [inline SofaKernel/modules/SofaCore/src/sofa/core/MultiVecId.h]: I'm not a fan of "hiding" the return type of functions. I think it messes with code readability.
Is it really necessary and/or useful?

Comment (damienmarchal) [reply SofaKernel/modules/SofaCore/src/sofa/core/MultiVecId.h]: Without the auto you need to keep the dependency to Data.

**Pull request closed without merging**